odoo/addons/test_new_api/tests/test_new_fields.py

```diff
@@ -929,24 +929,106 @@ def test_31_prefetch(self):
-        self.assertFalse(message.id)
+        # create a new message; fields have their default value if not given
+        new_msg = self.env['test_new_api.message'].new({'body': "XXX"})
+        self.assertFalse(new_msg.id)
+        self.assertEqual(new_msg.body, "XXX")
+        self.assertEqual(new_msg.author, user)
 
         # assign some fields; should have no side effect
-        message.discussion = discussion
-        message.body = BODY = "May the Force be with you."
-        self.assertEqual(message.discussion, discussion)
-        self.assertEqual(message.body, BODY)
-        self.assertFalse(message.author)
-        self.assertNotIn(message, discussion.messages)
+        new_msg.discussion = discussion
+        new_msg.body = "YYY"
+        self.assertEqual(new_msg.discussion, discussion)
+        self.assertEqual(new_msg.body, "YYY")
+        self.assertNotIn(new_msg, discussion.messages)
 
         # check computed values of fields
-        self.assertEqual(message.name, "[%s] %s" % (discussion.name, ''))
-        self.assertEqual(message.size, len(BODY))
+        self.assertEqual(new_msg.name, "[%s] %s" % (discussion.name, user.name))
+        self.assertEqual(new_msg.size, 3)
+
+    def test_40_new_one2many(self):
+        """ Test new records with one2many fields. """
+        # create a new discussion
+        msg0 = self.env['test_new_api.message'].create({'body': "XXX"})
+        msg1 = self.env['test_new_api.message'].create({'body': "WWW"})
+        new_disc = self.env['test_new_api.discussion'].new({
+            'name': "Stuff",
+            'messages': [
+                (4, msg0.id),
+                (4, msg1.id),
+                (1, msg1.id, {'body': "YYY"}),
+                (0, 0, {'body': "ZZZ"})
+            ],
+        })
+        self.assertFalse(new_disc.id)
+        self.assertEqual(len(new_disc.messages), 3)
+
+        # the one2many lines are new lines
+        new_msg0, new_msg1, new_msg2 = new_disc.messages
+        self.assertFalse(new_msg0.id)
+        self.assertFalse(new_msg1.id)
+        self.assertFalse(new_msg2.id)
+
+        # the one2many has its inverse field set
+        self.assertEqual(new_msg0.discussion, new_disc)
+        self.assertEqual(new_msg1.discussion, new_disc)
+        self.assertEqual(new_msg2.discussion, new_disc)
+
+        # new lines are connected to their origin
+        self.assertEqual(new_msg0._origin, msg0)
+        self.assertEqual(new_msg1._origin, msg1)
+        self.assertFalse(new_msg2._origin)
+
+        # the field values are either specific, or the same as the origin
+        self.assertEqual(new_msg0.body, "XXX")
+        self.assertEqual(new_msg1.body, "YYY")
+        self.assertEqual(new_msg2.body, "ZZZ")
+
+        # original records have not been modified
+        self.assertFalse(msg0.discussion)
+        self.assertFalse(msg1.discussion)
+
+        self.assertEqual(msg0.body, "XXX")
+        self.assertEqual(msg1.body, "WWW")
```

Comment:
Add tests for accessing all kinds of relational fields (many2one, one2many, many2many) with all special cases.